File: mrgeo-core/src/main/java/org/mrgeo/mapalgebra/RasterMapOp.java
/*
 * Copyright 2009-2014 DigitalGlobe, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations under the License.
 */

package org.mrgeo.mapalgebra;

import org.mrgeo.aggregators.Aggregator;
import org.mrgeo.aggregators.MeanAggregator;
import org.mrgeo.aggregators.ModeAggregator;
import org.mrgeo.buildpyramid.BuildPyramidSpark;
import org.mrgeo.image.MrsImagePyramid;
import org.mrgeo.image.MrsImagePyramidMetadata;
import org.mrgeo.image.MrsImagePyramidMetadata.Classification;
import org.mrgeo.mapreduce.job.JobCancelledException;
import org.mrgeo.mapreduce.job.JobFailedException;
import org.mrgeo.progress.Progress;
import org.mrgeo.data.DataProviderFactory;
import org.mrgeo.data.DataProviderFactory.AccessMode;
import org.mrgeo.data.image.MrsImageDataProvider;
import org.mrgeo.utils.HadoopUtils;

import java.awt.image.RenderedImage;
import java.io.IOException;

/**
 * Serves as a base class for all MapOp's that return a raster. Note that subclasses
 * should also determine whether or not to implement interfaces used during the
 * processing of map algebra, including BoundsCalculator, InputsCalculator,
 * MaximumZoomLevelCalculator, TileClusterInfoCalculator, and TileSizeCalculator.
 * Please refer to javadocs for those interfaces to determine if a subclassed MapOp
 * should implement them.
 */
public abstract class RasterMapOp extends MapOp implements OutputProducer
{
  /**
   * Creates a new empty pyramid and puts it on the temporary file list.
   * 
   * @param image
   * @return
   * @throws IOException 
   */
  public static MrsImagePyramid flushRasterMapOpOutput(MapOp op, int argumentNumber)
      throws IOException, JobFailedException, JobCancelledException
  {
    if (!(op instanceof RasterMapOp))
    {
      throw new IllegalArgumentException("Expected raster input data for argument " +
          argumentNumber);
    }

    RasterMapOp rasterOp = (RasterMapOp)op;
    // if our source is a file, then use it directly.
    if (rasterOp.getOutputName() == null)
    {
      throw new IllegalArgumentException("Invalid raster input data - no resource name " +
          "for argument " + argumentNumber);
    }
    return MrsImagePyramid.open(rasterOp.getOutputName(), op.getProviderProperties());
  }

  protected RenderedImage _output;
  protected String _outputName = null;

  /**
   * Returns the output object w/o blocking. This must be called after determineOutput has
   * completed.
   * 
   * @return
   * @throws IOException 
   */
  public RenderedImage getRasterOutput() throws IOException
  {
    return _output;
  }

  @Override
  public void postBuild(Progress p, boolean buildPyramid) throws IOException,
  JobFailedException, JobCancelledException
  {
    if (buildPyramid)
    {
      MrsImagePyramid pyramid = MrsImagePyramid.open(_outputName, getProviderProperties());
      MrsImagePyramidMetadata metadata = pyramid.getMetadata();

      Aggregator aggregator;
      if (metadata.getClassification() == Classification.Continuous)
      {
        aggregator = new MeanAggregator();
      }
      else
      {
        aggregator = new ModeAggregator();
      }
      BuildPyramidSpark.build(_outputName, aggregator, createConfiguration(), getProviderProperties());
    }
    p.complete();
  }

  /**
   * Returns the output resource name. This is _only_ relevant if the output has
   * already been calculated or it exists on disk for some other reason. If it doesn't exist and
   * needs to be calculated this method should return null. This is for optimization purposes only,
   * do not assume that this will be non-null.
   */
  @Override
  public String getOutputName()
  {
    return _outputName;
  }

  @Override
  public void setOutputName(final String output)
  {
    _outputName = output;
  }

  public String resolveOutputName() throws IOException
  {
    if (_outputName == null)
    {
      MrsImageDataProvider dp = DataProviderFactory.getMrsImageDataProvider(
          HadoopUtils.createRandomString(40),
          AccessMode.OVERWRITE,
          getProviderProperties());
      _outputName = dp.getResourceName();
      addTempResource(_outputName);
    }
    return _outputName;
  }

  /**
   * After a map op chain is executed, moveOutput will be called on the root map op. By default, the
   * map op's output is stored in a location other than where its final resting place will be, and
   * this method is responsible for moving the content to that location (e.g. toName).
   * 
   * @param fs
   * @param toName
   * @throws IOException
   */
  @Override
  public void moveOutput(final String toName) throws IOException
  {
    // make sure the toName doesn't exist, otherwise the move() will move the output into a directory
    // under the toName.
    DataProviderFactory.delete(toName, getProviderProperties());

    MrsImageDataProvider dp = DataProviderFactory.getMrsImageDataProvider(getOutputName(),
        AccessMode.READ, getProviderProperties());
    if (dp != null)
    {
      dp.move(toName);
    }
    _outputName = toName;
  }

  /**
   * Returns a clone of this MapOp and all its children. This will not include any intermediate
   * results or temporary variables involved in computation.
   */
  @Override
  public RasterMapOp clone()
  {
    final RasterMapOp result = (RasterMapOp) super.clone();
    result._output = null;
    return result;
  }
}

File: mrgeo-core/src/main/scala/org/mrgeo/mapalgebra/raster/RasterMapOp.scala
/*
 * Copyright 2009-2016 DigitalGlobe, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations under the License.
 *
 */

package org.mrgeo.mapalgebra.raster

import java.io.IOException

import org.apache.spark.SparkContext
import org.mrgeo.data.DataProviderFactory.AccessMode
import org.mrgeo.data.image.MrsImageDataProvider
import org.mrgeo.data.raster.RasterWritable
import org.mrgeo.data.rdd.RasterRDD
import org.mrgeo.data.tile.TileIdWritable
import org.mrgeo.data.{DataProviderFactory, ProviderProperties}
import org.mrgeo.image.MrsPyramidMetadata
import org.mrgeo.mapalgebra.MapOp
import org.mrgeo.mapalgebra.parser.{ParserException, ParserFunctionNode, ParserNode, ParserVariableNode}
//import org.mrgeo.utils.MrGeoImplicits._
import org.mrgeo.utils.tms.{TileBounds, TMSUtils}
import org.mrgeo.utils.{GDALUtils, SparkUtils}

object RasterMapOp {

  val EPSILON: Double = 1e-8

  def isNodata(value:Double, nodata:Double):Boolean = {
    if (nodata.isNaN) {
      value.isNaN
    }
    else {
      nodata == value
    }
  }
  def isNotNodata(value:Double, nodata:Double):Boolean = {
    if (nodata.isNaN) {
      !value.isNaN
    }
    else {
      nodata != value
    }
  }

  def nearZero(v:Double):Boolean = {
    if (v >= -EPSILON && v <= EPSILON) {
      true
    }
    else {
      false
    }
  }

  def decodeToRaster(node:ParserNode, variables: String => Option[ParserNode]): Option[RasterMapOp] = {
    node match {
    case func: ParserFunctionNode => func.getMapOp match {
    case raster: RasterMapOp => Some(raster)
    case _ => throw new ParserException("Term \"" + node + "\" is not a raster input")
    }
    case variable: ParserVariableNode =>
      MapOp.decodeVariable(variable, variables).getOrElse(throw new ParserException("Variable \"" + node + " has not been assigned")) match {
      case func: ParserFunctionNode => func.getMapOp match {
      case raster: RasterMapOp => Some(raster)
      case _ => throw new ParserException("Term \"" + node + "\" is not a raster input")
      }
      case _ => throw new ParserException("Term \"" + node + "\" is not a raster input")
      }
    case _ => throw new ParserException("Term \"" + node + "\" is not a raster input")
    }
  }

  def createEmptyRasterRDD(context: SparkContext, tb: TileBounds, zoom: Int) = {
    val tileBuilder = Array.newBuilder[(TileIdWritable, RasterWritable)]
    for (ty <- tb.s to tb.n) {
      for (tx <- tb.w to tb.e) {
        val id = TMSUtils.tileid(tx, ty, zoom)

        val tuple = (new TileIdWritable(id), new RasterWritable())
        tileBuilder += tuple
      }
    }

    new RasterRDD(context.parallelize(tileBuilder.result()))
  }
}


abstract class RasterMapOp extends MapOp {

  private var meta:MrsPyramidMetadata = null


  def rdd():Option[RasterRDD]

  def rdd(zoom: Int):Option[RasterRDD] = {
    if (meta != null && zoom == meta.getMaxZoomLevel) {
      rdd()
    }
    else {
      throw new IOException("rdd(zoom) rdd with zoom level other than max zoom not supported for raster RDD")
    }
  }

  def metadata():Option[MrsPyramidMetadata] =  Option(meta)
  def metadata(meta:MrsPyramidMetadata) = { this.meta = meta}

  def save(output: String, providerProperties:ProviderProperties, context:SparkContext) = {
    rdd() match {
    case Some(rdd) =>
      val provider: MrsImageDataProvider =
        DataProviderFactory.getMrsImageDataProvider(output, AccessMode.OVERWRITE, providerProperties)
      metadata() match {
      case Some(metadata) =>
        val meta = metadata

        SparkUtils.saveMrsPyramid(rdd, provider, meta, meta.getMaxZoomLevel,
          context.hadoopConfiguration, providerproperties =  providerProperties)
      case _ =>
        throw new IOException("Unable to save - no metadata was assigned in " + this.getClass.getName)
      }
    case _ =>
      throw new IOException("Unable to save - no RDD was assigned in " + this.getClass.getName)
    }
  }

  def toRaster(exact:Boolean = false) = {
    val rasterrdd = rdd() getOrElse(throw new IOException("Can't load RDD! Ouch! " + getClass.getName))
    SparkUtils.mergeTiles(rasterrdd, meta.getMaxZoomLevel, meta.getTilesize, meta.getDefaultValues,
      if (exact) meta.getBounds else null)
  }

  def toDataset(exact:Boolean = false) = {
    val rasterrdd = rdd() getOrElse(throw new IOException("Can't load RDD! Ouch! " + getClass.getName))

    val zoom = meta.getMaxZoomLevel
    val tilesize = meta.getTilesize

    val raster = SparkUtils.mergeTiles(rasterrdd, zoom, tilesize, meta.getDefaultValues,
      if (exact) meta.getBounds else null)

    val bounds = if (exact) {
      meta.getBounds
    }
    else {
      TMSUtils.tileBounds(meta.getBounds, zoom, tilesize)
    }

    GDALUtils.toDataset(raster, meta.getDefaultValue(0), bounds)

  }
}
